src/tty-ask-password-agent/tty-ask-password-agent.c

```diff
@@ -209,6 +228,72 @@ static int ask_password_plymouth(
         return 0;
 }
 
+static void free_consoles(struct console *con, unsigned int num) {
+        unsigned int n;
+        if (!con || !num)
+                return;
+        for (n = 0; n < num; n++)
+                free(con[n].tty);
+        free(con);
+}
+
+static const char *current_dev = "/dev/console";
+static int collect_consoles(struct console **consoles, unsigned int *num) {
+        _cleanup_free_ char *active = NULL;
+        const char *word, *state;
+        struct console *con = NULL;
+        size_t con_len = 0, len;
+        unsigned int count = 0;
+        int ret;
+
+        ret = read_one_line_file("/sys/class/tty/console/active", &active);
+        if (ret < 0)
+                return ret;
+        FOREACH_WORD(word, len, active, state) {
+                _cleanup_free_ char *tty = NULL;
+
+                if (strneq(word, "tty0", len) &&
+                    read_one_line_file("/sys/class/tty/tty0/active", &tty) >= 0) {
+                        word = tty;
+                        len = strlen(tty);
+                }
+                con = greedy_realloc((void**)&con, &con_len, 1+count, sizeof(struct console));
+                if (!con) {
+                        log_oom();
+                        return -errno;
+                }
+                if (asprintf(&con[count].tty, "/dev/%.*s", (int)len, word) < 0) {
+                        free_consoles(con, count);
+                        log_oom();
+                        return -errno;
+                }
+                con[count].pid = 0;
+                count++;
+        }
+        if (!con) {
+                con = greedy_realloc((void**)&con, &con_len, 1, sizeof(struct console));
+                if (!con) {
+                        log_oom();
+                        return -errno;
+                }
+                con[0].tty = strdup(current_dev);
+                if (!con[0].tty) {
+                        free_consoles(con, 1);
+                        log_oom();
+                        return -errno;
+                }
+                con[0].pid = 0;
+                count++;
+        }
+
+        if (num)
+                *num = count;
+        if (consoles)
+                *consoles = con;
```

Comment:
There's no point in making that optional, especially as this function has exactly one user. `assert()` on those variables to be non-null early in this function.